File: tl/src/main/java/com/github/badoualy/telegram/tl/api/account/TLPasswordInputSettings.java
package com.github.badoualy.telegram.tl.api.account;

import com.github.badoualy.telegram.tl.TLContext;
import com.github.badoualy.telegram.tl.core.TLBytes;
import com.github.badoualy.telegram.tl.core.TLObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

import static com.github.badoualy.telegram.tl.StreamUtils.readInt;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLBytes;
import static com.github.badoualy.telegram.tl.StreamUtils.readTLString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeInt;
import static com.github.badoualy.telegram.tl.StreamUtils.writeString;
import static com.github.badoualy.telegram.tl.StreamUtils.writeTLBytes;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID;
import static com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize;
import static com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize;

/**
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
public class TLPasswordInputSettings extends TLObject {

    public static final int CONSTRUCTOR_ID = 0x86916deb;

    protected int flags;

    protected TLBytes newSalt;

    protected TLBytes newPasswordHash;

    protected String hint;

    protected String email;

    private final String _constructor = "account.passwordInputSettings#86916deb";

    public TLPasswordInputSettings() {
    }

    public TLPasswordInputSettings(TLBytes newSalt, TLBytes newPasswordHash, String hint, String email) {
        this.newSalt = newSalt;
        this.newPasswordHash = newPasswordHash;
        this.hint = hint;
        this.email = email;
    }

    private void computeFlags() {
        flags = 0;
        flags = newSalt != null ? (flags | 1) : (flags & ~1);
        flags = newPasswordHash != null ? (flags | 1) : (flags & ~1);
        flags = hint != null ? (flags | 1) : (flags & ~1);
        flags = email != null ? (flags | 2) : (flags & ~2);
    }

    @Override
    public void serializeBody(OutputStream stream) throws IOException {
        computeFlags();

        writeInt(flags, stream);
        if ((flags & 1) != 0) {
            if (newSalt == null) throwNullFieldException("newSalt", flags);
            writeTLBytes(newSalt, stream);
        }
        if ((flags & 1) != 0) {
            if (newPasswordHash == null) throwNullFieldException("newPasswordHash", flags);
            writeTLBytes(newPasswordHash, stream);
        }
        if ((flags & 1) != 0) {
            if (hint == null) throwNullFieldException("hint", flags);
            writeString(hint, stream);
        }
        if ((flags & 2) != 0) {
            if (email == null) throwNullFieldException("email", flags);
            writeString(email, stream);
        }
    }

    @Override
    @SuppressWarnings({"unchecked", "SimplifiableConditionalExpression"})
    public void deserializeBody(InputStream stream, TLContext context) throws IOException {
        flags = readInt(stream);
        newSalt = (flags & 1) != 0 ? readTLBytes(stream, context) : null;
        newPasswordHash = (flags & 1) != 0 ? readTLBytes(stream, context) : null;
        hint = (flags & 1) != 0 ? readTLString(stream) : null;
        email = (flags & 2) != 0 ? readTLString(stream) : null;
    }

    @Override
    public int computeSerializedSize() {
        computeFlags();

        int size = SIZE_CONSTRUCTOR_ID;
        size += SIZE_INT32;
        if ((flags & 1) != 0) {
            if (newSalt == null) throwNullFieldException("newSalt", flags);
            size += computeTLBytesSerializedSize(newSalt);
        }
        if ((flags & 1) != 0) {
            if (newPasswordHash == null) throwNullFieldException("newPasswordHash", flags);
            size += computeTLBytesSerializedSize(newPasswordHash);
        }
        if ((flags & 1) != 0) {
            if (hint == null) throwNullFieldException("hint", flags);
            size += computeTLStringSerializedSize(hint);
        }
        if ((flags & 2) != 0) {
            if (email == null) throwNullFieldException("email", flags);
            size += computeTLStringSerializedSize(email);
        }
        return size;
    }

    @Override
    public String toString() {
        return _constructor;
    }

    @Override
    public int getConstructorId() {
        return CONSTRUCTOR_ID;
    }

    public TLBytes getNewSalt() {
        return newSalt;
    }

    public void setNewSalt(TLBytes newSalt) {
        this.newSalt = newSalt;
    }

    public TLBytes getNewPasswordHash() {
        return newPasswordHash;
    }

    public void setNewPasswordHash(TLBytes newPasswordHash) {
        this.newPasswordHash = newPasswordHash;
    }

    public String getHint() {
        return hint;
    }

    public void setHint(String hint) {
        this.hint = hint;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }
}

File: tl/src/main/kotlin/com/github/badoualy/telegram/tl/api/account/TLPasswordInputSettings.kt
package com.github.badoualy.telegram.tl.api.account

import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_BOOLEAN
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_CONSTRUCTOR_ID
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_DOUBLE
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT32
import com.github.badoualy.telegram.tl.TLObjectUtils.SIZE_INT64
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLBytesSerializedSize
import com.github.badoualy.telegram.tl.TLObjectUtils.computeTLStringSerializedSize
import com.github.badoualy.telegram.tl.core.TLBytes
import com.github.badoualy.telegram.tl.core.TLObject
import com.github.badoualy.telegram.tl.serialization.TLDeserializer
import com.github.badoualy.telegram.tl.serialization.TLSerializer
import java.io.IOException
import kotlin.Any
import kotlin.Boolean
import kotlin.Int
import kotlin.Long
import kotlin.String
import kotlin.jvm.Throws

/**
 * account.passwordInputSettings#21ffa60d
 *
 * @author Yannick Badoual yann.badoual@gmail.com
 * @see <a href="http://github.com/badoualy/kotlogram">http://github.com/badoualy/kotlogram</a>
 */
class TLPasswordInputSettings() : TLObject() {
    var newSalt: TLBytes? = null

    var newPasswordHash: TLBytes? = null

    var hint: String? = null

    var email: String? = null

    var newSecureSalt: TLBytes? = null

    var newSecureSecret: TLBytes? = null

    var newSecureSecretId: Long? = null

    private val _constructor: String = "account.passwordInputSettings#21ffa60d"

    override val constructorId: Int = CONSTRUCTOR_ID

    constructor(
            newSalt: TLBytes?,
            newPasswordHash: TLBytes?,
            hint: String?,
            email: String?,
            newSecureSalt: TLBytes?,
            newSecureSecret: TLBytes?,
            newSecureSecretId: Long?
    ) : this() {
        this.newSalt = newSalt
        this.newPasswordHash = newPasswordHash
        this.hint = hint
        this.email = email
        this.newSecureSalt = newSecureSalt
        this.newSecureSecret = newSecureSecret
        this.newSecureSecretId = newSecureSecretId
    }

    protected override fun computeFlags() {
        _flags = 0
        updateFlags(newSalt, 1)
        updateFlags(newPasswordHash, 1)
        updateFlags(hint, 1)
        updateFlags(email, 2)
        updateFlags(newSecureSalt, 4)
        updateFlags(newSecureSecret, 4)
        updateFlags(newSecureSecretId, 4)
    }

    @Throws(IOException::class)
    override fun serializeBody(tlSerializer: TLSerializer) = with (tlSerializer)  {
        computeFlags()

        writeInt(_flags)
        doIfMask(newSalt, 1) { writeTLBytes(it) }
        doIfMask(newPasswordHash, 1) { writeTLBytes(it) }
        doIfMask(hint, 1) { writeString(it) }
        doIfMask(email, 2) { writeString(it) }
        doIfMask(newSecureSalt, 4) { writeTLBytes(it) }
        doIfMask(newSecureSecret, 4) { writeTLBytes(it) }
        doIfMask(newSecureSecretId, 4) { writeLong(it) }
    }

    @Throws(IOException::class)
    override fun deserializeBody(tlDeserializer: TLDeserializer) = with (tlDeserializer)  {
        _flags = readInt()
        newSalt = readIfMask(1) { readTLBytes() }
        newPasswordHash = readIfMask(1) { readTLBytes() }
        hint = readIfMask(1) { readString() }
        email = readIfMask(2) { readString() }
        newSecureSalt = readIfMask(4) { readTLBytes() }
        newSecureSecret = readIfMask(4) { readTLBytes() }
        newSecureSecretId = readIfMask(4) { readLong() }
    }

    override fun computeSerializedSize(): Int {
        computeFlags()

        var size = SIZE_CONSTRUCTOR_ID
        size += SIZE_INT32
        size += getIntIfMask(newSalt, 1) { computeTLBytesSerializedSize(it) }
        size += getIntIfMask(newPasswordHash, 1) { computeTLBytesSerializedSize(it) }
        size += getIntIfMask(hint, 1) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(email, 2) { computeTLStringSerializedSize(it) }
        size += getIntIfMask(newSecureSalt, 4) { computeTLBytesSerializedSize(it) }
        size += getIntIfMask(newSecureSecret, 4) { computeTLBytesSerializedSize(it) }
        size += getIntIfMask(newSecureSecretId, 4) { SIZE_INT64 }
        return size
    }

    override fun toString() = _constructor

    override fun equals(other: Any?): Boolean {
        if (other !is TLPasswordInputSettings) return false
        if (other === this) return true

        return _flags == other._flags
                && newSalt == other.newSalt
                && newPasswordHash == other.newPasswordHash
                && hint == other.hint
                && email == other.email
                && newSecureSalt == other.newSecureSalt
                && newSecureSecret == other.newSecureSecret
                && newSecureSecretId == other.newSecureSecretId
    }
    companion object  {
        const val CONSTRUCTOR_ID: Int = 0x21ffa60d.toInt()
    }
}
